### PR TITLE
Lazily load fonts that are embedded within the application

### PR DESCRIPTION
**Pull request by jason-simmons**
Previously the engine would load all embedded fonts listed in the app's
font manifest during startup.  This change creates a Skia font manager that
is backed by the engine's AssetManager and can load embedded font assets
on demand.

### CODE REVIEW

**Comment by jason-simmons** [comment]
@cbracken @chinmaygarde

File: lib/ui/text/asset_manager_font_provider.cc [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    : asset_manager_(asset_manager) {}
+
+AssetManagerFontProvider::~AssetManagerFontProvider() = default;
+
+size_t AssetManagerFontProvider::GetFamilyCount() {
+  return family_names_.size();
+}
+
+std::string AssetManagerFontProvider::GetFamilyName(int index) {
+  return family_names_[index];
```
Comment (chinmaygarde) [inline lib/ui/text/asset_manager_font_provider.cc]: Make this method const (we own the superclass I think) and access the name with `find` on the collection. We don't want to end up inserting items into the collection unnecessarily.

File: lib/ui/text/asset_manager_font_provider.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+SkFontStyleSet* AssetManagerFontProvider::MatchFamily(
```
Comment (chinmaygarde) [inline lib/ui/text/asset_manager_font_provider.cc]: Nit: I know we don't do this consistently, but can we start putting the superclass this adds an override for in a comment above similar to how we decorate the headers? IMO, it makes things much clearer.

File: lib/ui/text/asset_manager_font_provider.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      return nullptr;
+
+    std::vector<uint8_t>* asset_buf_ptr = asset_buf.release();
+    sk_sp<SkData> asset_data = SkData::MakeWithProc(
+        asset_buf_ptr->data(), asset_buf_ptr->size(), VectorReleaseProc,
+        asset_buf_ptr);
+    std::unique_ptr<SkMemoryStream> stream = SkMemoryStream::Make(asset_data);
+
+    // Ownership of the stream is transferred.
+    asset.typeface = SkTypeface::MakeFromStream(stream.release());
```
Comment (chinmaygarde) [inline lib/ui/text/asset_manager_font_provider.cc]: This ctor takes an index, are you sure your don't have to pass the index onto the typeface?

TBH, I am not entirely clear how this mechanism with the indices works.

File: lib/ui/text/asset_manager_font_provider.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  SkTypeface* createTypeface(int index) override;
+
+  // |SkFontStyleSet|
+  SkTypeface* matchStyle(const SkFontStyle& pattern) override;
+
+ private:
+  fml::RefPtr<blink::AssetManager> asset_manager_;
+
+  struct TypefaceAsset {
+    TypefaceAsset(std::string a) : asset(a) {}
```
Comment (chinmaygarde) [inline lib/ui/text/asset_manager_font_provider.h]: move?

File: third_party/txt/BUILD.gn [hunk 5]
```diff
@@ -133,6 +130,19 @@ txt_common_executable_deps = [
   "$flutter_root/fml",  # For ICU initialization.
 ]
 
+source_set("txt_test_utils") {
+  sources = [
+    "tests/utils.cc",
+    "tests/utils.h",
```
Comment (chinmaygarde) [inline third_party/txt/BUILD.gn]: Nit: Maybe call this txt_test_utils.h|cc to better disambiguate utility files from multiple subcomponents.

File: third_party/txt/tests/utils.cc [hunk 6]
```diff
@@ -40,4 +46,74 @@ void SetCommandLine(fxl::CommandLine cmd) {
   gCommandLine = std::move(cmd);
 }
 
+void RegisterFontsFromPath(TypefaceFontAssetProvider& font_provider,
+                           std::string directory_path) {
+#if defined(_WIN32)
```
Comment (chinmaygarde) [inline third_party/txt/tests/utils.cc]: I know you just moved these files, but we should replace the platform specific file handling routines with the ones in FML.

File: lib/ui/text/asset_manager_font_provider.cc (hunk 1 above)
Comment (jason-simmons) [reply lib/ui/text/asset_manager_font_provider.cc]: done (family_names_ is a vector - no risk of unintended insertion)

File: lib/ui/text/asset_manager_font_provider.cc (hunk 2 above)
Comment (jason-simmons) [reply lib/ui/text/asset_manager_font_provider.cc]: done

File: lib/ui/text/asset_manager_font_provider.cc (hunk 3 above)
Comment (jason-simmons) [reply lib/ui/text/asset_manager_font_provider.cc]: The index just identifies the font within this font set, ranging from 0 through count() - 1

File: lib/ui/text/asset_manager_font_provider.h (hunk 4 above)
Comment (jason-simmons) [reply lib/ui/text/asset_manager_font_provider.h]: done

File: third_party/txt/BUILD.gn (hunk 5 above)
Comment (jason-simmons) [reply third_party/txt/BUILD.gn]: done

File: third_party/txt/tests/utils.cc (hunk 6 above)
Comment (jason-simmons) [reply third_party/txt/tests/utils.cc]: Does FML offer a way to list files within a directory?

File: lib/ui/text/asset_manager_font_provider.cc (hunk 1 above)
Comment (chinmaygarde) [reply lib/ui/text/asset_manager_font_provider.cc]: Oh. In that case, a bounds check is necessary.

File: lib/ui/text/asset_manager_font_provider.cc (hunk 3 above)
Comment (chinmaygarde) [reply lib/ui/text/asset_manager_font_provider.cc]: Gotcha

File: third_party/txt/tests/utils.cc (hunk 6 above)
Comment (chinmaygarde) [reply third_party/txt/tests/utils.cc]: If it doesn't, we should add it there (in a different patch is fine) so that all platforms can use it.

**Pull request merged**